Expand 3x3 to NxN tic-tac-toe

ndsl7109256 opened the pull request:
Just simply modify **board_size** to change the size of the game.
I generalize the check function of winning and scoring function.

Commit:
Expand 3x3 to NxN tic-tac-toe

jserv commented on ttt.c:
```diff
@@ -176,22 +279,27 @@ int get_input(char player)
 int main()
 {
-    char table[] = {' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '};
+    char *table = malloc(sizeof(char) * grid_num);
```
Release the allocated memory when `table` is not used anymore.

jserv commented on ttt.c:
```diff
@@ -39,41 +43,92 @@ void draw_board(const char *t)
     printf("\n");
 }
 
+char check_row(const char *t)
```
Since `check_row` and `check_col` have lines in common, can you introduce a helper function which accepts a flag to indicate wether row or column should be checked?

jserv commented on ttt.c:
```diff
@@ -3,9 +3,15 @@
 #include <stdlib.h>
 #include <string.h>
 
+#define board_size 4
```
Capitalize the macro identifiers. That is, use `BOARD_SIZE` rather than `board_size`.

jserv commented on ttt.c:
```diff
@@ -3,9 +3,15 @@
 
+#define board_size 4
+#define grid_num (board_size) * (board_size)
+#define get_index(i, j) (i) * (board_size) + (j)
```
Ditto. Be aware of capitalized macro identifers.

jserv commented on ttt.c:
```diff
@@ -3,9 +3,15 @@
 void draw_board(const char *t)
 {
-    const int M = 3, N = 3;
+    const int M = board_size, N = board_size;
```
Drop the use of `M` and `N`. Use `BOARD_SIZE` directly.

jserv commented on ttt.c:
```diff
@@ -39,41 +45,93 @@ void draw_board(const char *t)
 
+char check_row_col(const char *t, int flag)
+{
+    int i_shift[2], j_shift[2];
+    if (flag == ROW) {
+        i_shift[0] = 0;
+        i_shift[1] = 0;
+        j_shift[0] = 1;
+        j_shift[1] = 2;
+    } else if (flag == COL) {
+        i_shift[0] = 1;
+        i_shift[1] = 2;
+        j_shift[0] = 0;
+        j_shift[1] = 0;
+    }
```
Simplify these lines by using `memcpy`.

jserv commented on ttt.c:
```diff
@@ -3,9 +3,15 @@
 #include <string.h>
 
+#define board_size 4
+#define grid_num (board_size) * (board_size)
```
Rename to `N_GRIDS`.

jserv commented on ttt.c:
```diff
@@ -3,20 +3,22 @@
-#define get_index(i, j) (i) * (board_size) + (j)
+#define BOARD_SIZE 4
+#define N_GRIDS (BOARD_SIZE) * (BOARD_SIZE)
+#define get_index(i, j) (i) * (BOARD_SIZE) + (j)
```
Capitalize this macro.

jserv commented on ttt.c:
```diff
@@ -30,50 +38,94 @@ void draw_board(const char *t)
+    char c = check_row_col(t, COL);
+    if (c != ' ')
+        return c;
+    char md = check_diagonal(t, MAIN);
```
The parameters of `check_diagonal`, namely `MAIN` and `BACK`, are ambiguous, making them hard to distinguish. Can you improve?

ndsl7109256 replied:
Refering https://en.wikipedia.org/wiki/Main_diagonal
Would `PRIMARY` and `SECONDARY` be a greater name?

jserv replied:
> Would PRIMARY and SECONDARY be a greater name?

It sounds good.

jserv reviewed (changes requested):
Rebase the latest `master` branch.

ndsl7109256 commented:
I extend `print_moves` to NxN also and color the steps!

jserv reviewed (changes requested):
Use `git rebase` instead of `git merge`.

jserv reviewed (commented):
Decouple the colorized changes into another pull request.

ndsl7109256 committed:
Decouple the colorized change

jserv commented:
I noticed a surprising delay between initiating my first move and the AI displaying its first movement. Following the alteration of `BOARD_SIZE` from 4 to 5, this delay has noticeably increased. Could you verify this?

ndsl7109256 commented:
I thought it's because of the dramatically increasing size of the search tree(We have to solve decision tree with height $O(N^2)$ and each node has $O(N^2)$ child). In fact, I got the same problem from 3 to 4 before introducing alpha-beta pruning way.

Maybe a easy solution is to limit the max search depth. Like
```c
#define MAX_DEPTH 6
...
int negamax(char *table, int depth, char player, int alpha, int beta)
{
     if(depth == MAX_DEPTH){
         return get_score(table, player);
     }
     ...
}

```
I try it in `BOARD_SIZE` = 5 with `MAX_DEPTH` = 6, and the AI still behaves well.

ndsl7109256 commented:
Another way to improve might be limit the search grid. We could focus the grid nearby the grid has 'O' or 'X'. (It is obviously not a good strategy to go a step far away from the other 'O', 'X' in big board). However, it requires more code change and I thought limiting the search depth is good enough for tic-tac-toe, maybe we could implement limiting search depth first?

jserv commented:
> I thought it's because of the dramatically increasing size of the search tree(We have to solve decision tree with height O(N2) and each node has O(N2) child). In fact, I got the same problem from 3 to 4 before introducing alpha-beta pruning way.
> Maybe a easy solution is to limit the max search depth. Like

@visitorckw, can you comment this?

visitorckw commented:
Limiting the search depth is a fairly intuitive approach, but I'm unsure if the effectiveness is good enough. The n, m, k game has a long and storied history, and it has already been proven that in situations where the board size is sufficiently large (for example, 13x13) and k = 5, the player who moves first is guaranteed to win [1]. I believe that in the case of k = 3, compared to k = 5, it should be simpler to implement a winning strategy for the first player by utilizing local search techniques, such as threat space search[2]. Of course, this could potentially become an issue to be addressed in a later pull request.

Link: http://fragrieu.free.fr/SearchingForSolutions.pdf [1]
Link: https://ir.nctu.edu.tw/bitstream/11536/50437/1/552101.pdf [2]

visitorckw commented:
Another interesting approach to consider is adopting Monte Carlo Tree Search. Regardless of its effectiveness, this method allows for better control over the AI's decision-making time.

jserv reviewed (changes requested):
Rebase the latest `main` branch.

ndsl7109256 committed:
Limit the search depth with MAX_SEARCH_DEPTH

To avoid the increasing size of the search tree, we limit the depth of search tree to control the searching time.

jserv commented:
Thank @ndsl7109256 for contributing!
